lib/protobuf/wire/varint.ex

```diff
@@ -0,0 +1,208 @@
+      quote(do: x0 + bsl(x1, 7))
+    },
+    {
+      quote(do: <<1::1, x0::7, 1::1, x1::7, 0::1, x2::7>>),
+      quote(do: x0 + bsl(x1, 7) + bsl(x2, 14))
+    },
+    {
+      quote(do: <<1::1, x0::7, 1::1, x1::7, 1::1, x2::7, 0::1, x3::7>>),
+      quote(do: x0 + bsl(x1, 7) + bsl(x2, 14) + bsl(x3, 21))
+    },
+    {
+      quote(do: <<1::1, x0::7, 1::1, x1::7, 1::1, x2::7, 1::1, x3::7, 0::1, x4::7>>),
+      quote(do: x0 + bsl(x1, 7) + bsl(x2, 14) + bsl(x3, 21) + bsl(x4, 28))
+    },
+    {
+      quote do
+        <<1::1, x0::7, 1::1, x1::7, 1::1, x2::7, 1::1, x3::7, 1::1, x4::7, 0::1, x5::7>>
+      end,
+      quote do
+        x0 +
+          bsl(x1, 7) +
+          bsl(x2, 14) +
+          bsl(x3, 21) +
+          bsl(x4, 28) +
+          bsl(x5, 35)
+      end
+    },
+    {
+      quote do
+        <<1::1, x0::7, 1::1, x1::7, 1::1, x2::7, 1::1, x3::7, 1::1, x4::7, 1::1, x5::7, 0::1,
+          x6::7>>
+      end,
+      quote do
+        x0 +
+          bsl(x1, 7) +
+          bsl(x2, 14) +
+          bsl(x3, 21) +
+          bsl(x4, 28) +
+          bsl(x5, 35) +
+          bsl(x6, 42)
+      end
+    },
+    {
+      quote do
+        <<1::1, x0::7, 1::1, x1::7, 1::1, x2::7, 1::1, x3::7, 1::1, x4::7, 1::1, x5::7, 1::1,
+          x6::7, 0::1, x7::7>>
+      end,
+      quote do
+        x0 +
+          bsl(x1, 7) +
+          bsl(x2, 14) +
+          bsl(x3, 21) +
+          bsl(x4, 28) +
+          bsl(x5, 35) +
+          bsl(x6, 42) +
+          bsl(x7, 49)
+      end
+    },
+    {
+      quote do
+        <<1::1, x0::7, 1::1, x1::7, 1::1, x2::7, 1::1, x3::7, 1::1, x4::7, 1::1, x5::7, 1::1,
+          x6::7, 1::1, x7::7, 0::1, x8::7>>
+      end,
+      quote do
+        x0 +
+          bsl(x1, 7) +
+          bsl(x2, 14) +
+          bsl(x3, 21) +
+          bsl(x4, 28) +
+          bsl(x5, 35) +
+          bsl(x6, 42) +
+          bsl(x7, 49) +
+          bsl(x8, 56)
+      end
+    },
+    {
+      quote do
+        <<1::1, x0::7, 1::1, x1::7, 1::1, x2::7, 1::1, x3::7, 1::1, x4::7, 1::1, x5::7, 1::1,
+          x6::7, 1::1, x7::7, 1::1, x8::7, 0::1, x9::7>>
+      end,
+      quote do
+        band(
+          x0 +
+            bsl(x1, 7) +
+            bsl(x2, 14) +
+            bsl(x3, 21) +
+            bsl(x4, 28) +
+            bsl(x5, 35) +
+            bsl(x6, 42) +
+            bsl(x7, 49) +
+            bsl(x8, 56) +
+            bsl(x9, 63),
+          unquote(@mask64)
+        )
+      end
+    }
+  ]
+
+  defmacro decoder(kind, name, args \\ [], do: block) do
+    def_success(kind, name, args, block) ++ [def_failure(kind, name, args)]
+  end
+
+  defp def_success(kind, name, args, block) do
+    args = Enum.map(args, fn arg -> {arg, [line: 1], nil} end)
+
+    for {pattern, expression} <- @varints do
+      head = quote(do: unquote(name)(<<unquote(pattern), rest::bits>>, unquote_splicing(args)))
+
+      body =
+        quote do
+          var!(value) = unquote(expression)
```

Comment:
I will need some help here, metaprogramming is definitely not my forte 😅

I feel there is a better way of doing this. Right now my `skip_varint` function will trigger a compilation warning, because it does not use the `value` variable in its body. I wonder what is the correct way of fixing that.